### ISSUE TITLE
python package eval AUC above 1.0

### ISSUE THREAD

**Issue by galbraun** [dmlc/xgboost#7492]
When I train a model using the python package when I print the eval, after some rounds the roc-auc get above 1.0 :

> [48]	validation_0-auc:0.99959	validation_1-auc:0.97537
[49]	validation_0-auc:0.99961	validation_1-auc:0.97529
[50]	validation_0-auc:0.99970	validation_1-auc:0.97569
[51]	validation_0-auc:0.99982	validation_1-auc:0.97636
[52]	validation_0-auc:0.99984	validation_1-auc:0.97646
[53]	validation_0-auc:0.99991	validation_1-auc:0.97675
[54]	validation_0-auc:0.99994	validation_1-auc:0.97686
[55]	validation_0-auc:0.99995	validation_1-auc:0.97701
[56]	validation_0-auc:1.00003	validation_1-auc:0.97713
[57]	validation_0-auc:1.00008	validation_1-auc:0.97738
[58]	validation_0-auc:1.00013	validation_1-auc:0.97739
[59]	validation_0-auc:1.00018	validation_1-auc:0.97746
[60]	validation_0-auc:1.00022	validation_1-auc:0.97749
[61]	validation_0-auc:1.00023	validation_1-auc:0.97757
[62]	validation_0-auc:1.00024	validation_1-auc:0.97749
[63]	validation_0-auc:1.00027	validation_1-auc:0.97750
[64]	validation_0-auc:1.00029	validation_1-auc:0.97753
[65]	validation_0-auc:1.00031	validation_1-auc:0.97755
[66]	validation_0-auc:1.00032	validation_1-auc:0.97750
[67]	validation_0-auc:1.00033	validation_1-auc:0.97738
[68]	validation_0-auc:1.00035	validation_1-auc:0.97765
[69]	validation_0-auc:1.00036	validation_1-auc:0.97757
[70]	validation_0-auc:1.00038	validation_1-auc:0.97750

validation_0 is the training set and validation_1 is the validation set.

Env: 
- os: Amazon Linux
- python version: 3.9.7
- xgboost release: 1.5.0

**Comment by hcho3** [comment]
Is it possible to share your data and training script?

**Comment by galbraun** [comment]
I could reproduce the problem with this code:

> from xgboost import XGBClassifier
from sklearn.datasets import make_classification
X,y = make_classification(n_samples=10000, n_features=1000)
X_val,y_val = make_classification(n_samples=10000, n_features=1000)
xgboost_params = {'eta': 0.3, 'gamma': 3.930975040805966e-06, 'max_depth': 10, 'min_child_weight': 1, 'subsample': 0.9, 'lambda': 0.018522161601581322, 'alpha': 0.5440050106523926, 'scale_pos_weight': 47, 'colsample_bytree': 0.8, 'colsample_bylevel': 0.8}
xgb = xgboost.XGBClassifier(n_jobs=-1, n_estimators=100, **xgboost_params)
xgb.fit(X, y, eval_metric='auc', eval_set=[(X, y), (X_val, y_val)])

and got this output:

> [0]	validation_0-auc:0.50840	validation_1-auc:0.50080
[1]	validation_0-auc:0.97599	validation_1-auc:0.49689
[2]	validation_0-auc:0.97927	validation_1-auc:0.49666
[3]	validation_0-auc:0.98135	validation_1-auc:0.49612
[4]	validation_0-auc:0.98255	validation_1-auc:0.49517
[5]	validation_0-auc:0.98701	validation_1-auc:0.49121
[6]	validation_0-auc:0.99136	validation_1-auc:0.49462
[7]	validation_0-auc:0.99233	validation_1-auc:0.49322
[8]	validation_0-auc:0.99308	validation_1-auc:0.49374
[9]	validation_0-auc:0.99534	validation_1-auc:0.49198
[10]	validation_0-auc:0.99597	validation_1-auc:0.49196
[11]	validation_0-auc:0.99616	validation_1-auc:0.49209
[12]	validation_0-auc:0.99740	validation_1-auc:0.49170
[13]	validation_0-auc:0.99814	validation_1-auc:0.49235
[14]	validation_0-auc:0.99893	validation_1-auc:0.49193
[15]	validation_0-auc:0.99943	validation_1-auc:0.49196
[16]	validation_0-auc:0.99951	validation_1-auc:0.49228
[17]	validation_0-auc:0.99968	validation_1-auc:0.49286
[18]	validation_0-auc:0.99981	validation_1-auc:0.49211
[19]	validation_0-auc:0.99994	validation_1-auc:0.49220
[20]	validation_0-auc:1.00001	validation_1-auc:0.49195
[21]	validation_0-auc:1.00003	validation_1-auc:0.49207
[22]	validation_0-auc:1.00005	validation_1-auc:0.49209
[23]	validation_0-auc:1.00006	validation_1-auc:0.49180
[24]	validation_0-auc:1.00006	validation_1-auc:0.49213
[25]	validation_0-auc:1.00006	validation_1-auc:0.49225
[26]	validation_0-auc:1.00007	validation_1-auc:0.49209
[27]	validation_0-auc:1.00007	validation_1-auc:0.49222
[28]	validation_0-auc:1.00007	validation_1-auc:0.49212
[29]	validation_0-auc:1.00007	validation_1-auc:0.49249
[30]	validation_0-auc:1.00007	validation_1-auc:0.49271

**Comment by trivialfis** [comment]
Thanks for opening the issue.  It's caused by the floating-point error.  The latest XGBoost uses double instead of float for evaluation metrics and the issue is fixed.

**Comment by trivialfis** [comment]
```
[91]	validation_0-auc:1.00000	validation_1-auc:0.50168
[92]	validation_0-auc:1.00000	validation_1-auc:0.50169
[93]	validation_0-auc:1.00000	validation_1-auc:0.50172
[94]	validation_0-auc:1.00000	validation_1-auc:0.50174
[95]	validation_0-auc:1.00000	validation_1-auc:0.50171
[96]	validation_0-auc:1.00000	validation_1-auc:0.50172
[97]	validation_0-auc:1.00000	validation_1-auc:0.50175
[98]	validation_0-auc:1.00000	validation_1-auc:0.50171
[99]	validation_0-auc:1.00000	validation_1-auc:0.50170

```

**Comment by galbraun** [comment]
@trivialfis in which version? I still get AUC above 1.0 :
![image](https://user-images.githubusercontent.com/9768714/145705880-6d84c43f-7e16-4b24-9e2f-29f04401a9db.png)

**Comment by trivialfis** [comment]
It's on the master branch, hasn't been released yet.
